test/dynamics/pulse/test_pulse_to_signals.py

```diff
@@ -303,15 +320,79 @@ def test_multiple_channels_with_gaps(self):
         samples2 = np.append(np.zeros(15), np.append(gauss_samples, np.zeros(10)))
         samples3 = np.append(np.zeros(20), np.append(gauss_samples, np.zeros(5)))
 
-        self.assertAllClose(signals[0].samples, samples0, atol=1e-14, rtol=1e-14)
-        self.assertAllClose(signals[1].samples, samples1, atol=1e-14, rtol=1e-14)
-        self.assertAllClose(signals[2].samples, samples2, atol=1e-14, rtol=1e-14)
-        self.assertAllClose(signals[3].samples, samples3, atol=1e-14, rtol=1e-14)
+        self.assertAllClose(signals[0].samples, samples0, atol=1e-7, rtol=1e-7)
+        self.assertAllClose(signals[1].samples, samples1, atol=1e-7, rtol=1e-7)
+        self.assertAllClose(signals[2].samples, samples2, atol=1e-7, rtol=1e-7)
+        self.assertAllClose(signals[3].samples, samples3, atol=1e-7, rtol=1e-7)
         self.assertTrue(signals[0].carrier_freq == 5.0)
         self.assertTrue(signals[1].carrier_freq == 3.1)
         self.assertTrue(signals[2].carrier_freq == 0.0)
         self.assertTrue(signals[3].carrier_freq == 4.0)
 
+    def test_get_samples(self):
+        """Test get samples of Pulse not get_waveform but get_samples function."""
+        gauss_get_waveform_samples = (
+            pulse.Gaussian(duration=5, amp=0.983, sigma=2.0).get_waveform().samples
+        )
+        gauss_get_samples = get_samples(Gaussian(duration=5, amp=0.983, sigma=2.0))
+        self.assertTrue(isinstance(gauss_get_samples, np.ndarray))
+        self.assertAllClose(gauss_get_samples, gauss_get_waveform_samples, atol=1e-7, rtol=1e-7)
+
+    def test_lru_cache_expr(self):
+        """Test lru_cache of lru_cache_expr function."""
+        gauss_envelop = Gaussian(duration=5, amp=0.983, sigma=2.0).envelope
+        self.assertTrue(
+            _lru_cache_expr(gauss_envelop, Array.default_backend())
+            is _lru_cache_expr(gauss_envelop, Array.default_backend())
+        )
+
+
+class TestJaxGetSamples(QiskitDynamicsTestCase, TestJaxBase):
```

Comment:
Similar to the preceding comment, I don't think it's necessary to have a separate class for checking the behaviour of `get_samples`. These tests could be moved to the previous class, by rewriting them to verify that the converter returns the same thing, whether you pass in a `Waveform` or a `SymbolicPulse`. 

The `jit` test is great and very important, but again, it can be moved to the previous class and rewritten to work with the converter.

Reply:
For this I mean to fully get rid of this test class, and to move the cases it's testing into the preceding test class. by calling `InstructionToSignals` directly.

Reply:
The class you said moving to is `TestPulseToSignals`?
If I move to the class, we fail to pass the test because `TestPulseToSignals` is not a subclass of `TestJaxBase`.
On the other hand, if `TestPulseToSignals` inherits `TestJaxBase`, all the test in `TestPulseToSignals` will be skipped in python test and executed in only JAX test.

Reply:
I understand now. Maybe just rename this class then to `TestPulseToSignalsJAXTransformations` or something along these lines.

Reply:
rename and fix at https://github.com/Qiskit/qiskit-dynamics/pull/149/commits/454f1194463df8cfe5bf002343e81fa92fdc04b4 and https://github.com/Qiskit/qiskit-dynamics/pull/149/commits/2814cf0587ceabc6354cbf628560536bcad24577